### DISCUSSION TITLE
Javascript Event Listener

### DISCUSSION THREAD

**Discussion by WSH032** [zauberzeug/nicegui#2383] [Ideas / Feature Requests]
I would like to add an event listener that runs only in JavaScript rather than in Python.

After studying Vue and `nicegui.templates.html`, I've found the following hack method:

```py
from nicegui import ui


_js = """\
Vue.withModifiers((evt) => {
    alert(evt);
}, ['prevent'])
"""


label = ui.label("Click me")
label._props[":onClick"] = _js
label.update()

if __name__ == '__main__':
    ui.run(reload=False)
```

But, how about adding a api, like:

```py
_js = "(evt) => { alert(evt); }"

label.on("click.prevent", js_handler=_js)
```

It can be simplified by leveraging the existing code in nicegui:

https://github.com/zauberzeug/nicegui/blob/ccab1812fd414080f0f844827b2f7d8ea1daf77f/nicegui/templates/index.html#L145-L179

**Comment by falkoschindler** [comment]
Hi @WSH032, thanks for the interesting feature request!
This is certainly a quite advanced feature and we need to carefully think about the API and how/where to implement it.
Do you already have an implementation in mind? Maybe you can create a draft pull request. 🙂

**Reply by WSH032** [reply]
Hi @falkoschindler , thank you for looking into this!

I currently don't have the time to delve into the testing workflow of `nicegui`, so I might not be able to create a pull request that aligns well with `CONTRIBUTION.md`.

Here's a quick demo, looking forward to your feedback.

https://github.com/WSH032/nicegui/commit/644859e3bd2caccc0e25fc60deb4e017553e35ac

```py
if __name__ == "__main__":
    from nicegui import ui

    ui.input("press `enter` key").on("click.once", js_handler="(evt) => alert(evt)")

    ui.run(reload=False)
```

![demo](https://github.com/zauberzeug/nicegui/assets/126865849/f1e3c6a5-c503-4b16-a2c8-91d0a38c4710)

**EDIT:**

Please ignore `press enter key`

Initially, I intended to use `keyup.enter` for demonstration, but in the end, I forgot to change the label 🤷‍♂️

**Reply by falkoschindler** [reply]
Sorry for the late reply, @WSH032! Your proposal looks really nice! I think this would already make a great pull request. We can add tests and documentation for you. 🙂
My only concern is that `args` and throttling parameters are simply ignored without any warning when using `js_handler`. But we can probably fix that by [overloading](https://docs.python.org/3/library/typing.html#typing.overload) the `.on()` method.

**Reply by WSH032** [reply]
That's ok. I've already created a pull request #2536 .

**Comment by falkoschindler** [comment]
This feature has just been released in version [1.4.16](https://github.com/zauberzeug/nicegui/releases/tag/v1.4.16). 🎉
